# BERT Embeeding Natural Language Processing </a>

## K Nearest Neighbors Model for a Classification Problem: Classify Product Reviews as Positive or Negative

In this notebook, we use the K Nearest Neighbors method to build a classifier to predict the __isPositive__ field of our review dataset (that is very similar to the final project dataset).


1. <a href="#1">Reading the dataset</a>
2. <a href="#2">Exploratory data analysis</a>
3. <a href="#3">Text Processing: Stop words removal and stemming</a>
4. <a href="#4">Train - Validation Split</a>
5. <a href="#5">Data processing with Pipeline</a>
6. <a href="#6">Train the classifier</a>
7. <a href="#7">Test the classifier</a> Find more details on the KNN Classifier here: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html
8. <a href="#8">Ideas for improvement</a>

Overall dataset schema:
* __reviewText:__ Text of the review
* __summary:__ Summary of the review
* __verified:__ Whether the purchase was verified (True or False)
* __time:__ UNIX timestamp for the review
* __log_votes:__ Logarithm-adjusted votes log(1+votes). *This field is a processed version of the votes field. People can click on the "helpful" button when they find a customer review helpful. This increases the vote by 1. __log_votes__ is calculated like this log(1+votes). This formulation helps us get a smaller range for votes.*
* __isPositive:__ Whether the review is positive or negative (1 or 0)


## 1. <a name="1">Reading the dataset</a>
(<a href="#0">Go to top</a>)

We will use the __pandas__ library to read our dataset.

In [11]:
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/aws-samples/aws-machine-learning-university-accelerated-nlp/master/data/examples/AMAZON-REVIEW-DATA-CLASSIFICATION.csv')

print('The shape of the dataset is:', df.shape)

The shape of the dataset is: (70000, 6)


In [ ]:
# IMDB Dataset
#df = pd.read_csv('https://raw.githubusercontent.com/aws-samples/aws-machine-learning-university-accelerated-nlp/master/data/final_project/imdb_train.csv', header=0)

#train_df = pd.read_csv('https://raw.githubusercontent.com/aws-samples/aws-machine-learning-university-accelerated-nlp/master/data/final_project/imdb_train.csv', header=0)
#train_df.head()

#test_df = pd.read_csv('https://raw.githubusercontent.com/aws-samples/aws-machine-learning-university-accelerated-nlp/master/data/final_project/imdb_test.csv', header=0)
#test_df.head()


Let's look at the first 10 rows of the dataset.

In [ ]:
df.head(10)

reviewText  \
0  PURCHASED FOR YOUNGSTER WHO\nINHERITED MY "TOO...   
1                              unable to open or use   
2   Waste of money!!! It wouldn't load to my system.   
3  I attempted to install this OS on two differen...   
4  I've spent 14 fruitless hours over the past tw...   
5  I purchased the home and business because I wa...   
6  The download doesn't take long at all. And it'...   
7  This program is positively wonderful for word ...   
8   Fantastic protection!!  Great customer support!!   
9  Obviously Win 7 now the last great operating s...   

                                             summary  verified        time  \
0                                IDEAL FOR BEGINNER!      True  1361836800   
1                                          Two Stars      True  1452643200   
2                                       Dont buy it!      True  1433289600   
3  I attempted to install this OS on two differen...      True  1518912000   
4                                   Do NOT Download.      True  1441929600   
5        Quicken home and business  not for amatures      True  1335312000   
6                                             Great!      True  1377993600   
7                             Terrific for practice.     False  1158364800   
8                                         Five Stars      True  1478476800   
9                                         Five Stars      True  1471478400   

   log_votes  isPositive  
0   0.000000         1.0  
1   0.000000         0.0  
2   0.000000         0.0  
3   0.000000         0.0  
4   1.098612         0.0  
5   0.000000         0.0  
6   0.000000         1.0  
7   2.397895         1.0  
8   0.000000         1.0  
9   0.000000         1.0

## 2. <a name="2">Exploratory data analysis</a>
(<a href="#0">Go to top</a>)

Let's look at the distribution of __isPositive__ field.

In [ ]:
df["isPositive"].value_counts()

1.0    43692
0.0    26308
Name: isPositive, dtype: int64

We can check the number of missing values for each columm below.

In [ ]:
print(df.isna().sum())

reviewText    11
summary       14
verified       0
time           0
log_votes      0
isPositive     0
dtype: int64


We have missing values in our text fields.

## 3. <a name="3">Text Processing: Stop words removal and stemming</a>
(<a href="#0">Go to top</a>)

In [13]:
df=df.dropna()
print(df.isna().sum())

reviewText    0
summary       0
verified      0
time          0
log_votes     0
isPositive    0
dtype: int64


In [12]:
!pip install transformers torch scikit-learn


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.7/7.7 MB 46.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.0/302.0 kB 31.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 105.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 77.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.0/295.0 kB 38.3 MB/s eta 0:00:00


## 4. <a name="4">Train - Validation Split</a>
(<a href="#0">Go to top</a>)

Let's split our dataset into training (90%) and validation (10%).

In [14]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Split the dataset into train and test sets
train_data, test_data, train_labels, test_labels = train_test_split(df["reviewText"], df['isPositive'], test_size=0.1, random_state=42)

## Use BERT for text embeddings:
(<a href="#0">Go to top</a>)

You can use the Hugging Face Transformers library to load a pre-trained BERT model and tokenize your text dat

In [15]:
import torch
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

cuda


In [17]:
from transformers import BertTokenizer, BertModel
import torch
from tqdm import tqdm

# Move the model to GPU if available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
# BERT-based Classifier
# Load BERT tokenizer and model, move to GPU

#tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
#bert_model = BertModel.from_pretrained("bert-base-uncased").to(device)

# xlm-roberta-base :
from transformers import AutoTokenizer, AutoModelForMaskedLM

tokenizer = AutoTokenizer.from_pretrained('xlm-roberta-base')
model = AutoModelForMaskedLM.from_pretrained("xlm-roberta-base").to(device)


# Tokenize and encode the text data, move to GPU
max_length = 128
X_train_tokens = tokenizer(list(train_data), truncation=True, padding=True, max_length=max_length, return_tensors="pt", add_special_tokens=True).to(device)
X_test_tokens = tokenizer(list(test_data), truncation=True, padding=True, max_length=max_length, return_tensors="pt", add_special_tokens=True).to(device)

# Calculate BERT embeddings for the text data
def get_bert_embeddings(tokens):
    embeddings = []
    for i in tqdm(range(len(tokens['input_ids']))):
        with torch.no_grad():
            output = model(input_ids=tokens['input_ids'][i].unsqueeze(0), attention_mask=tokens['attention_mask'][i].unsqueeze(0))
        embeddings.append(output[0].squeeze().mean(dim=0).cpu().numpy())
    return embeddings

X_train_bert_embeddings = get_bert_embeddings(X_train_tokens)
X_test_bert_embeddings = get_bert_embeddings(X_test_tokens)

KeyboardInterrupt: ignored

## 6. <a name="6">Train the classifier</a>
(<a href="#0">Go to top</a>)

We train our classifier with __.fit()__ on our training dataset.
Train a KNN model:
Now that you have BERT embeddings for your text data, you can train a KNN model using scikit-learn.

This code demonstrates how to use BERT for text embeddings and then train a KNN model for sentiment analysis of AWS product reviews. Make sure to replace 'your_dataset.csv' with the actual path to your dataset file and adjust other parameters as needed. You may also fine-tune the model and preprocessing steps to improve performance.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Initialize and train the KNN classifier
knn = KNeighborsClassifier(n_neighbors=10)
knn.fit(X_train_bert_embeddings, train_labels)

NB = GaussianNB()
NB.fit(X_train_bert_embeddings, train_labels)

# Train a classifier on BERT embeddings (you can use any classifier of your choice)
# Here, we'll use Logistic Regression as an example
lr = LogisticRegression(max_iter=1000)
lr.fit(X_train_bert_embeddings, train_labels)

#rf=RandomForestClassifier()
#rf.fit(train_embeddings, train_labels)
#xgb=GradientBoostingClassifier()
#xgb.fit(train_embeddings, train_labels)



LogisticRegression(max_iter=1000)

##Test the classifier
(Go to top)

To evaluate the KNN model's performance on sentiment classification, you can generate a classification report and a confusion matrix. Here's how you can do it using scikit-learn:
|--|--|--|

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix

# Make predictions on the test data
predictions = knn.predict(X_test_bert_embeddings)

# Calculate accuracy
accuracy = accuracy_score(test_labels, predictions)
print(f'KNN Accuracy: {accuracy * 100:.2f}%')
#predictions = rf.predict(test_embeddings)
predictions = NB.predict(X_test_bert_embeddings)
accuracy = accuracy_score(test_labels, predictions)
print(f'Naive Bais Accuracy: {accuracy * 100:.2f}%')

#predictions = rf.predict(test_embeddings)
predictions = lr.predict(X_test_bert_embeddings)
accuracy = accuracy_score(test_labels, predictions)
print(f'Logistic Regression Accuracy: {accuracy * 100:.2f}%')

KNN Accuracy: 84.41%
Naive Bais Accuracy: 66.36%
Logistic Regression Accuracy: 88.03%


In [ ]:
from sklearn.metrics import classification_report, confusion_matrix

# Make predictions on the test data
predictions = knn.predict(X_test_bert_embeddings)
# Create a classification report
class_report = classification_report(test_labels, predictions, target_names=['negative', 'positive'])

# Create a confusion matrix
conf_matrix = confusion_matrix(test_labels, predictions)

# Print the classification report and confusion matrix
print("Classification Report:")
print(class_report)

print("\nConfusion Matrix:")
print(conf_matrix)

#predictions = rf.predict(test_embeddings)
predictions = NB.predict(X_test_bert_embeddings)
# Create a classification report
class_report = classification_report(test_labels, predictions, target_names=['negative', 'positive'])

# Create a confusion matrix
conf_matrix = confusion_matrix(test_labels, predictions)

# Print the classification report and confusion matrix
print("Classification Report:")
print(class_report)

print("\nConfusion Matrix:")
print(conf_matrix)

#predictions = rf.predict(test_embeddings)
predictions = lr.predict(X_test_bert_embeddings)
# Create a classification report
class_report = classification_report(test_labels, predictions, target_names=['negative', 'positive'])

# Create a confusion matrix
conf_matrix = confusion_matrix(test_labels, predictions)

# Print the classification report and confusion matrix
print("Classification Report:")
print(class_report)

print("\nConfusion Matrix:")
print(conf_matrix)




Classification Report:
              precision    recall  f1-score   support

    negative       0.78      0.81      0.79      2583
    positive       0.88      0.87      0.88      4415

    accuracy                           0.84      6998
   macro avg       0.83      0.84      0.83      6998
weighted avg       0.85      0.84      0.84      6998


Confusion Matrix:
[[2082  501]
 [ 590 3825]]
Classification Report:
              precision    recall  f1-score   support

    negative       0.53      0.74      0.62      2583
    positive       0.80      0.62      0.70      4415

    accuracy                           0.66      6998
   macro avg       0.67      0.68      0.66      6998
weighted avg       0.70      0.66      0.67      6998


Confusion Matrix:
[[1918  665]
 [1689 2726]]
Classification Report:
              precision    recall  f1-score   support

    negative       0.84      0.84      0.84      2583
    positive       0.91      0.90      0.90      4415

    accuracy         

***Bag of words Method :***

Binary | Max Features | KNN (n_neighbors) | precision (macro avg) | recall (macro avg) | f1-score (macro avg) | precision (weighted avg) | recall (weighted avg) | f1-score (weighted avg) | Accuracy
------|-----|----|------|------|------|------|------|------|--------
True  | 5   | 3  | 0.50 | 0.50 | 0.49 | 0.50 | 0.50 | 0.49 | 0.49932
True  | 10  | 3  | 0.52 | 0.52 | 0.52 | 0.52 | 0.52 | 0.52 | 0.51768
True  | 25  | 3  | 0.56 | 0.56 | 0.56 | 0.56 | 0.56 | 0.56 | 0.55660
True  | 75  | 3  | 0.62 | 0.62 | 0.62 | 0.62 | 0.62 | 0.62 | 0.62376
True  | 100 | 3  | 0.63 | 0.63 | 0.63 | 0.63 | 0.63 | 0.63 | 0.63080
True  | 5   | 5  | 0.52 | 0.52 | 0.50 | 0.52 | 0.52 | 0.50 | 0.52140
True  | 10  | 5  | 0.52 | 0.52 | 0.52 | 0.52 | 0.52 | 0.52 | 0.52288
True  | 25  | 5  | 0.56 | 0.56 | 0.56 | 0.56 | 0.56 | 0.56 | 0.56472
True  | 75  | 5  | 0.64 | 0.64 | 0.64 | 0.64 | 0.64 | 0.64 | 0.63856
True  | 100 | 5  | 0.64 | 0.64 | 0.64 | 0.64 | 0.64 | 0.64 | 0.63816
True  | 5   | 9  | 0.53 | 0.53 | 0.53 | 0.53 | 0.53 | 0.53 | 0.53012
True  | 10  | 9  | 0.53 | 0.53 | 0.53 | 0.53 | 0.53 | 0.53 | 0.52932
True  | 25  | 9  | 0.57 | 0.57 | 0.57 | 0.57 | 0.57 | 0.57 | 0.57116
True  | 75  | 9  | 0.66 | 0.66 | 0.66 | 0.66 | 0.66 | 0.66 | 0.65572
True  | 100 | 9  | 0.66 | 0.66 | 0.66 | 0.66 | 0.66 | 0.66 | 0.66056
False | 5   | 3  | 0.51 | 0.51 | 0.51 | 0.51 | 0.51 | 0.51 | 0.50944
False | 10  | 3  | 0.52 | 0.52 | 0.52 | 0.52 | 0.52 | 0.52 | 0.52076
False | 25  | 3  | 0.59 | 0.59 | 0.59 | 0.59 | 0.59 | 0.59 | 0.59288
False | 75  | 3  | 0.62 | 0.62 | 0.62 | 0.62 | 0.62 | 0.62 | 0.61512
False | 100 | 3  | 0.62 | 0.62 | 0.62 | 0.62 | 0.62 | 0.62 | 0.61780
False | 5   | 5  | 0.52 | 0.52 | 0.52 | 0.52 | 0.52 | 0.52 | 0.51916
False | 10  | 5  | 0.52 | 0.52 | 0.52 | 0.52 | 0.52 | 0.52 | 0.52424
False | 25  | 5  | 0.60 | 0.60 | 0.60 | 0.60 | 0.60 | 0.60 | 0.60444
False | 75  | 5  | 0.63 | 0.63 | 0.63 | 0.63 | 0.63 | 0.63 | 0.63128
False | 100 | 5  | 0.64 | 0.64 | 0.64 | 0.64 | 0.64 | 0.64 | 0.63772
False | 5   | 9  | 0.53 | 0.53 | 0.53 | 0.53 | 0.53 | 0.53 | 0.52672
False | 10  | 9  | 0.53 | 0.53 | 0.53 | 0.53 | 0.53 | 0.53 | 0.53124
False | 25  | 9  | 0.62 | 0.62 | 0.61 | 0.62 | 0.62 | 0.61 | 0.61512
False | 75  | 9  | 0.65 | 0.65 | 0.65 | 0.65 | 0.65 | 0.65 | 0.64720
False | 100 | 9  | 0.66 | 0.66 | 0.66 | 0.66 | 0.66 | 0.66 | 0.65624
False | 100 | 15 | 0.67 | 0.67 | 0.67 | 0.67 | 0.67 | 0.67 | 0.66592

***BERT Transformers Method :***

KNeighborsClassifier :


x           |  x   |   x  |   x  |   x
------------|------|------|------|-----
accuracy    |      |      | 0.84 | 6998
macro avg   | 0.83 | 0.84 | 0.83 | 6998
weighted avg| 0.85 | 0.84 | 0.84 | 6998

GaussianNB :

x           |  x   |   x  |   x  |   x
------------|------|------|------|-----
accuracy    |      |      | 0.66 | 6998
macro avg   | 0.67 | 0.68 | 0.66 | 6998
weighted avg| 0.70 | 0.66 | 0.67 | 6998

LogisticRegression :

x           |  x   |   x  |   x  |   x
------------|------|------|------|-----
accuracy    |      |      | 0.88 | 6998
macro avg   | 0.87 | 0.87 | 0.87 | 6998
weighted avg| 0.88 | 0.88 | 0.88 | 6998


We notice that the accuracy are higher with BERT transformers than with bag of words.

***xlm-roberta-base Method :***

Testing environment :
Use with KNN = 10
LogisticRegression(max_iter=1000)

||Accuracy|
|-----|--------|
|Naive Bais Accuracy| 0.59 |
|KNN | 0.82 |
|Logistic Regression Accuracy| 0.89 |



In [20]:
from transformers import pipeline
classifier = pipeline("feature-extraction")
cl=classifier("I'm in Ireland for nine months for a year of study abroad.")
print(cl)

No model was supplied, defaulted to distilbert-base-cased and revision 935ac13 (https://huggingface.co/distilbert-base-cased).
Using a pipeline without specifying a model name and revision in production is not recommended.


[[[0.43934541940689087, 0.048983797430992126, -0.023671088740229607, -0.1146029606461525, -0.3084927201271057, -0.0031290289480239153, 0.3283122181892395, -0.13757012784481049, 0.08636888861656189, -1.0446380376815796, -0.19163751602172852, -0.011039575561881065, -0.04486796259880066, -0.17804713547229767, -0.5776118040084839, 0.11174338310956955, 0.18748050928115845, 0.334160715341568, -0.03570995479822159, -0.2652748227119446, 0.1237691193819046, -0.1384684294462204, 0.4667491018772125, -0.19610941410064697, 0.33075836300849915, 0.039786893874406815, 0.2870614528656006, 0.30507275462150574, -0.14777204394340515, 0.3422682285308838, 0.04050027206540108, 0.10515895485877991, -0.1593002825975418, 0.1220230907201767, -0.29520082473754883, 0.14641280472278595, 0.004145467188209295, -0.34395256638526917, -0.09908952564001083, -0.056648753583431244, -0.5108259320259094, 0.16445836424827576, 0.718429684638977, -0.2463853657245636, -0.1459101289510727, -0.38109198212623596, 0.0295606348663568